crates/astria-composer/src/executor/mod.rs

```diff
@@ -276,7 +287,122 @@ impl Executor {
                     }
                 }
             }
+        };
+
+        self.status
+            .send_modify(|status| status.is_connected = false);
+
+        // close the channel to avoid receiving any other txs before we drain the remaining
+        // sequence actions
+        self.serialized_rollup_transactions.close();
+
+        match &reason {
+            Ok(reason) => {
+                info!(reason, "shutting down");
+            }
+            Err(reason) => {
+                error!(%reason, "executor exited with error");
+                // we error out because of a failure to submit a bundle to the sequencer
+                // we do not want to proceed with the shutdown process in this case
+                return Err(eyre!(reason.to_string()));
+            }
+        };
+
+        let mut bundles_to_drain: VecDeque<SizedBundle> = VecDeque::new();
+        let mut bundles_drained = 0;
+
+        info!("draining sequence actions from the executor receiver channel");
+
+        // drain the receiver channel
+        while let Ok(seq_action) = self.serialized_rollup_transactions.try_recv() {
+            let rollup_id = seq_action.rollup_id;
+            if let Err(e) = bundle_factory.try_push(seq_action) {
+                warn!(
+                    rollup_id = %rollup_id,
+                    error = &e as &StdError,
+                    "failed to bundle sequence action, dropping it."
+                );
+            }
+        }
+
+        // when shutting down, drain all the remaining bundles and submit to the sequencer
+        // to avoid any bundle loss.
+        loop {
+            let bundle = bundle_factory.pop_now();
+            if bundle.is_empty() {
+                break;
+            }
+
+            bundles_to_drain.push_back(bundle);
+        }
+
+        let shutdown_logic = async {
+            // wait for the last bundle to be submitted
+            if !submission_fut.is_terminated() {
+                info!("waiting for the last bundle to be submitted to the sequencer");
+                match submission_fut.await {
+                    Ok(new_nonce) => {
+                        debug!(new_nonce = new_nonce, "drained bundle successfully");
+                        nonce = new_nonce;
+                    }
+                    Err(error) => {
+                        error!(%error, "failed submitting bundle to sequencer during shutdown; \
+                                aborting shutdown");
+                        return Err(error);
+                    }
+                }
+            }
+
+            info!(
+                no_of_bundles_to_drain = bundles_to_drain.len(),
+                "draining remaining bundles from bundle factory"
```

Comment:
don't mention `bundle factory`. It's an implementation detail that operators need not know.
```suggestion
                "submitting remaining transaction bundles to sequencer"
```